src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -408,60 +408,32 @@ public AgentRecord( Guid id, Process p, ITestAgent a, AgentStatus s )
         #region Nested Class - AgentDataBase
         /// <summary>
         ///  A simple class that tracks data about this
-        ///  agencies active and available agents
+        ///  agencies active and available agents.
+        ///  This class is required to be multi-thread safe.
         /// </summary>
         private class AgentDataBase
         {
-            private Dictionary<Guid, AgentRecord> agentData = new Dictionary<Guid, AgentRecord>();
+            private readonly Dictionary<Guid, AgentRecord> _agentData = new Dictionary<Guid, AgentRecord>();
+            private readonly object _lock = new object();
 
             public AgentRecord this[Guid id]
             {
-                get { return agentData[id]; }
-                set
-                {
-                    if ( value == null )
-                        agentData.Remove( id );
-                    else
-                        agentData[id] = value;
-                }
-            }
-
-            public AgentRecord this[ITestAgent agent]
-            {
                 get
                 {
-                    foreach( KeyValuePair<Guid, AgentRecord> entry in agentData)
+                    lock (_lock)
                     {
-                        AgentRecord r = entry.Value;
-                        if ( r.Agent == agent )
-                            return r;
+                        return _agentData[id];
                     }
-
-                    return null;
                 }
             }
 
-            public void Add( AgentRecord r )
+            public void Add(AgentRecord r)
             {
-                agentData[r.Id] = r;
-            }
-
-            public void Remove(Guid agentId)
-            {
-                agentData.Remove(agentId);
-            }
```

Comment:
I agree with the removal of unused methods, but I'm very surprised that this is unused. Seems like we should be removing entries when we stop the runner.

We've known about the problem of thread-safety for a long time - before the latest changes made it impossible to ignore. So it's possible that some of these method have calls that were deleted or commented out. Let's take a look at TestAgency to see if that's so.